CMakeLists.txt

```diff
@@ -182,13 +182,12 @@ elseif(ANDROID)
     list(APPEND PLATFORM_LIBRARIES log android)
 endif()
 
-if(UNIX AND NOT ANDROID)
-    # Use large-file APIs by default even on 32-bit platforms
-    # Does not work for Android: https://android.googlesource.com/platform/bionic/+/master/docs/32-bit-abi.md
-    check_type_size(off_t SIZEOF_OFF_T)
-    if(SIZEOF_OFF_T EQUAL "4")
-        add_compile_definitions("_FILE_OFFSET_BITS=64")
-    endif()
+if(UNIX)
```

Comment:
I still think we should be checking for glibc specifically here, because `UNIX` applies to Alpine (musl), Android (bionic), and macOS (libSystem), and AFAICT `_LARGEFILE*_SOURCE` is a glibc macro. Doesn't it make sense to keep this as narrow as possible?

Reply:
I've looked into this more, and it turns out everything about large file support on 32-bit platforms is a hot mess 😞 

In theory, I think _LARGEFILE_SOURCE and _FILE_OFFSET_BITS is supposed to be a universal unix thing. But not everyone complied with the spec. At the very least, it isn't just glibc. It is also used for uClibc (although they define the `__GLIBC__` macro, so 🤷‍♂️). It is also a thing [in musl](https://github.com/bminor/musl/blob/33338ebc853d37c80f0f236cc7a92cb0acc6aace/include/dirent.h#L59-L69), although it just redirects `readdir64` to `readdir` since all musl programs always use 64-bit file offsets ([their faq](https://wiki.musl-libc.org/faq.html#Q:-Do-I-need-to-define-%3Ccode%3E_LARGEFILE64_SOURCE%3C/code%3E-to-get-64bit-%3Ccode%3Eoff_t%3C/code%3E?) seems very angry about this subject...). Looking at how other build systems handle this, [meson defines `_FILE_OFFSET_BITS=64` everywhere except windows and darwin](https://github.com/mesonbuild/meson/blob/278c294aa45efc3e8b068bcd7632828ed5c92523/mesonbuild/compilers/compilers.py#L1208-L1227), and [LLVM defines `_LARGEFILE_SOURCE` and `_FILE_OFFSET_BITS=64`  on all 32-bit platforms except android<24](https://github.com/llvm/llvm-project/blob/master/llvm/cmake/modules/HandleLLVMOptions.cmake#L319-L326).

So it seems not harmful, and potentially helpful to define this widely, especially coupled with the check that `readdir64` is available. There are two other solutions, I've tried if we don't like this approach:

1. define `REALM_USE_STD_FILESYSTEM` on RPi builds, which calls into the filesystem code which is separately built with `_FILE_OFFSET_BITS=64` so does the right thing.
2. switch to using the i386 build of qemu-arm. This avoids running into the issue because it will use the 32-bit syscalls which don't return 64-bit values. However this would mean that our users would not be able to test their applications using a 64-bit qemu (which runs faster than the 32-bit one)

Do you prefer one of those?

Reply:
I was going to suggest just checking for glibc explicitly in CMake, but I see now it’s never that simple.
I’m okay with just leaving it as-is - I suppose it works, though hopefully it won’t bite us in the future.
I’m tempted by the `std::filesystem` option, but I’m afraid that might cause issues with backwards compatibility if we ever tried to target an older libstdc++, so I’d leave it for now.